llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp

```diff
@@ -227,13 +231,24 @@ static const unsigned MaxPHINumOperands = 128;
 /// avoids spending time checking the cost model and realizing that they will
 /// be inevitably scalarized.
 static bool isValidElementType(Type *Ty) {
+  // TODO: Support ScalableVectorType.
```

Comment:
Add the assertion for scalable types for now

Reply:
Add assert in `isValidElementType` will make SLP cannot work. But we can add assert inside `getNumElements`.

Reply:
Because `ScalableVectorType` is accidentally passed into `isValidElementType`.
```
diff --git a/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp b/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp
index 8925e3b2d91a..12bc491381fe 100644
--- a/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp
+++ b/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp
@@ -7512,7 +7512,7 @@ unsigned BoUpSLP::canMapToVector(Type *T) const {
     }
   }

-  if (!isValidElementType(EltTy))
+  if (isa<ScalableVectorType>(EltTy) || !isValidElementType(EltTy))
     return 0;
   uint64_t VTSize = DL->getTypeStoreSizeInBits(getWidenedType(EltTy, N));
   if (VTSize < MinVecRegSize || VTSize > MaxVecRegSize ||
@@ -18901,6 +18901,7 @@ bool SLPVectorizerPass::vectorizeChainsInBlock(BasicBlock *BB, BoUpSLP &R) {
       // No need to analyze deleted, vectorized and non-vectorizable
       // instructions.
       if (!VisitedInstrs.count(P) && !R.isDeleted(P) &&
+          !isa<ScalableVectorType>(P->getType()) &&
           isValidElementType(P->getType()))
         Incoming.push_back(P);
     }
@@ -19058,6 +19059,8 @@ bool SLPVectorizerPass::vectorizeChainsInBlock(BasicBlock *BB, BoUpSLP &R) {
       }
       if (TryToVectorizeRoot) {
         for (auto *V : It->operand_values()) {
+          if (isa<ScalableVectorType>(V->getType()))
+            continue;
           // Postponed instructions should not be vectorized here, delay their
           // vectorization.
           if (auto *VI = dyn_cast<Instruction>(V);
```
The first one is triggered by `llvm/test/Transforms/SLPVectorizer/RISCV/scalable-type-to-vect.ll`.
The second and the third one is triggered by `llvm/test/Transforms/SLPVectorizer/AArch64/scalable-vector.ll`.